tests/test_send_recv_two_workers.py

```diff
@@ -0,0 +1,185 @@
+
+        resp = await ep.recv_future()
+        obj = ucp.get_obj_from_msg(resp)
+        nframes, = struct.unpack("Q", obj[:8])  # first eight bytes for number of frames
+        gpu_frame_msg = obj[
+            8 : 8 + nframes
+        ]  # next nframes bytes for if they're GPU frames
+        is_gpus = struct.unpack("{}?".format(nframes), gpu_frame_msg)
+
+        sized_frame_msg = obj[8 + nframes :]  # then the rest for frame sizes
+        sizes = struct.unpack("{}Q".format(nframes), sized_frame_msg)
+
+        frames = []
+
+        for i, (is_gpu, size) in enumerate(zip(is_gpus, sizes)):
+            if size > 0:
+                resp = await ep.recv_obj(size, cuda=is_gpu)
+            else:
+                resp = await ep.recv_future()
+            frame = ucp.get_obj_from_msg(resp)
+            frames.append(frame)
+
+        header = pickle.loads(frames[0])
+        frames = frames[1:]  #
+        typ = pickle.loads(header["type"])
+        cuda_obj = typ.deserialize(header, frames)
+
+        print(cuda_obj, len(cuda_obj), type(cuda_obj))
+        asyncio.sleep(1)
+        # how should we confirm data is on two GPUs ?
+        await ep.send_obj(b"OK")
+        print("Shutting Down Client...")
+        ucp.destroy_ep(ep)
+        ucp.fin()
+
+    asyncio.get_event_loop().run_until_complete(read())
+
+
+async def generate_cuda_obj(cuda_type="cupy"):
+    """
+    generate cuda object and return serialized version
+    # if there is an error here it never bubbles up
+    """
+    with log_errors():
+        if cuda_type == "cupy":
+            import cupy
+            import distributed
+
+            data = cupy.arange(10)
+            frames = await distributed.comm.utils.to_frames(
+                data, serializers=("cuda", "pickle")
+            )
+        if cuda_type == "cudf":
+            import cudf
+            import numpy as np
+            asyncio.sleep(0.01)
+            # cdf = cudf.DataFrame({"a": range(5), "b": range(5)})
+            # cdf = cudf.DataFrame({"a": [1.0], "b": [1.0]}).head(0)
+            # cdf = cudf.DataFrame({"a": range(5), "b": range(5)}, index=[10,13,15,20,25])
+            size = 2**12
+            cdf = cudf.DataFrame({"a": np.random.random(size), "b": np.random.random(size)}, index=np.random.randint(size, size=size))
+            # cdf = cudf.DataFrame({"a": range(5), "b": [1.2, None, 2, 3, 4]}, index=[1, 2, 5, None, 6])
+            print(cdf.head().to_pandas())
+            header, _frames = cdf.serialize()
+            frames = [pickle.dumps(header)] + _frames
+
+    return frames
+
+
+def server(env, port):
+    # create listener receiver
+    # write cudf object
+    # confirm message is sent correctly
+
+    os.environ.update(env)
+    print("ENV SERVER: ", os.environ)
+
+    async def f(lisitener_port):
+        ucp.init()
+
+        # coroutine shows up when the client asks
+        # to connect
+        async def write(ep, li):
+            await asyncio.sleep(0.1)
```

Comment:
what purpose does the sleep operation serve here?

Reply:
This was designed to give a little bit of time for the client to startup.  Should we take it out ?